File: frontend/src/types/market.ts
export interface StockQuote {
  symbol: string;
  name: string;
  current_price: number;
  change: number;
  change_rate: number;
  change_sign: string; // 1=상한, 2=상승, 3=보합, 4=하한, 5=하락
  open_price: number;
  high_price: number;
  low_price: number;
  volume: number;
  trade_amount: number;
  prev_close: number;
  market_cap: number;
  per: number | null;
  pbr: number | null;
  eps: number | null;
}

export interface Candle {
  date: string;
  open: number;
  high: number;
  low: number;
  close: number;
  volume: number;
}

export interface OrderbookEntry {
  price: number;
  volume: number;
}

export interface Orderbook {
  symbol: string;
  ask: OrderbookEntry[];
  bid: OrderbookEntry[];
  total_ask_volume: number;
  total_bid_volume: number;
}

export interface TradeRecord {
  time: string;
  price: number;
  volume: number;
  change: number;
}

export interface IndexQuote {
  code: string;
  name: string;
  current: number;
  change: number;
  change_rate: number;
  change_sign: string;
}

export interface SearchResult {
  symbol: string;
  name: string;
  market: string;
}

export interface WatchlistItem {
  id: string;
  symbol: string;
  market: string;
  added_at: string;
  quote: StockQuote | null;
}

export interface Watchlist {
  id: string;
  name: string;
  created_at: string;
  items: WatchlistItem[];
}

// ── Order Types ──

export interface OrderCreate {
  symbol: string;
  side: "buy" | "sell";
  order_type: "limit" | "market";
  quantity: number;
  price?: number;
}

export interface OrderModify {
  quantity: number;
  price: number;
}

export interface OrderResponse {
  id: string;
  account_id: string;
  symbol: string;
  side: string;
  order_type: string;
  quantity: number;
  price: number | null;
  status: string;
  filled_quantity: number;
  filled_price: number | null;
  kis_order_id: string | null;
  submitted_at: string | null;
  created_at: string;
}

export interface PendingOrder {
  order_id: string;
  symbol: string;
  name: string;
  side: string;
  order_type: string;
  quantity: number;
  price: number;
  unfilled_qty: number;
  order_time: string;
}

export interface FilledOrder {
  order_id: string;
  symbol: string;
  name: string;
  side: string;
  quantity: number;
  price: number;
  total_amount: number;
  order_time: string;
  filled_time: string;
}

export interface BuyableAmount {
  orderable_cash: number;
  orderable_qty: number;
}

// ── Portfolio Types ──

export interface ForeignCurrencyBalance {
  currency: string;
  deposit: number;
  stock_value: number;
  total_value: number;
  exchange_rate: number;
}

export interface AccountBalance {
  total_value: number;
  cash: number;
  stock_value: number;
  total_pnl: number;
  total_pnl_rate: number;
  holding_count: number;
  foreign_balances: ForeignCurrencyBalance[];
  overseas_total_krw: number;
}

export interface Holding {
  symbol: string;
  name: string;
  quantity: number;
  avg_price: number;
  current_price: number;
  value: number;
  pnl: number;
  pnl_rate: number;
}

export interface OverseasHolding {
  symbol: string;
  name: string;
  market: string;
  currency: string;
  quantity: number;
  avg_price: number;
  current_price: number;
  value_foreign: number;
  pnl_foreign: number;
  pnl_rate: number;
  exchange_rate: number;
}

// ── Ranking Types ──

export interface RankItem {
  rank: number;
  symbol: string;
  name: string;
  current_price: number;
  change: number;
  change_rate: number;
  change_sign: string; // 1=상한 2=상승 3=보합 4=하한 5=하락
  volume: number;
  trade_amount: number;
}

export interface InvestorItem {
  investor: string;
  buy_volume: number;
  sell_volume: number;
  net_volume: number;
  buy_amount: number;
  sell_amount: number;
  net_amount: number;
}

// Color helpers for Korean market convention
export function getPriceColor(changeSign: string): string {
  // 1=상한, 2=상승 → red, 4=하한, 5=하락 → blue, 3=보합 → gray
  if (changeSign === "1" || changeSign === "2") return "text-red-500";
  if (changeSign === "4" || changeSign === "5") return "text-blue-500";
  return "text-gray-500";
}

export function getPriceSign(change: number): string {
  if (change > 0) return "+";
  return "";
}

File: backend/app/schemas/market.py
from pydantic import BaseModel, Field


class StockQuote(BaseModel):
    symbol: str
    name: str
    current_price: int
    change: int
    change_rate: float
    change_sign: str  # 1=상한, 2=상승, 3=보합, 4=하한, 5=하락
    open_price: int
    high_price: int
    low_price: int
    volume: int
    trade_amount: int
    prev_close: int
    market_cap: int = 0
    per: float | None = None
    pbr: float | None = None
    eps: int | None = None


class Candle(BaseModel):
    date: str  # YYYYMMDD or YYYYMMDDHHMM
    open: int
    high: int
    low: int
    close: int
    volume: int


class OrderbookEntry(BaseModel):
    price: int
    volume: int


class Orderbook(BaseModel):
    symbol: str
    ask: list[OrderbookEntry]  # 매도호가 (10단계, 1호가가 index 0)
    bid: list[OrderbookEntry]  # 매수호가 (10단계, 1호가가 index 0)
    total_ask_volume: int
    total_bid_volume: int


class TradeRecord(BaseModel):
    time: str
    price: int
    volume: int
    change: int


class IndexQuote(BaseModel):
    code: str
    name: str
    current: float
    change: float
    change_rate: float
    change_sign: str


class SearchResult(BaseModel):
    symbol: str
    name: str
    market: str  # KOSPI / KOSDAQ


class RankItem(BaseModel):
    """공통 랭킹 항목 — 거래량/등락률/시총/인기 등 동일 구조."""

    rank: int
    symbol: str
    name: str
    current_price: int
    change: int  # 전일대비
    change_rate: float  # 전일대비율(%)
    change_sign: str  # 1=상한 2=상승 3=보합 4=하한 5=하락
    volume: int  # 누적거래량
    trade_amount: int  # 누적거래대금


class InvestorItem(BaseModel):
    """투자자별 매매동향 항목."""

    investor: str  # 개인, 외국인, 기관계 등
    buy_volume: int
    sell_volume: int
    net_volume: int  # 순매수량
    buy_amount: int  # 매수금액 (백만원)
    sell_amount: int  # 매도금액 (백만원)
    net_amount: int  # 순매수금액 (백만원)
